src/metabase/query_processor/pivot/postprocess.clj

```diff
@@ -190,34 +190,36 @@
 (defn- build-headers
   "Combine row keys with column headers."
   [column-headers {:keys [pivot-cols pivot-rows column-titles]}]
-  (map (fn [h]
-         (if (and (seq pivot-cols) (not (seq pivot-rows)))
-           (concat (map #(get column-titles %) pivot-cols) h)
-           (concat (map #(get column-titles %) pivot-rows) h)))
```

Comment:
If there's one thing to learn from this PR, then it's this: **forget that `concat` exists** 😄. It's a function that almost never should be used, ever, ever. Instead, if you need to concatenate sequences together, use `into` with the target being a vector.

Overall, you should remove and avoid as much laziness as possible: https://clojure-goes-fast.com/blog/clojures-deadly-sin/

Reply:
Huge fan of this. I don't always love `mapv` though. It's easy to miss that this is used so the resulting type is a vector. If 4 months someone might come along and add some filter, some subsequent map, etc, and if that mapv is no longer last, we've changed how _everything_ in this sequence is accumulated. For this reason i much prefer an `(into [] ...)` or a simple call to `vec` around the whole so that it is clear that the return type is intended to be a vector. Both for performance reasons and _correctness_.

I did a `crit/quick-bench` on your version and got "Execution time mean : 6.391627 µs". If i just slap an `into` on the existing i get "Execution time mean : 6.267785 µs":

```clojure
(defn- build-headers
  "Combine row keys with column headers."
  [column-headers {:keys [pivot-cols pivot-rows column-titles] :as x}]
  (into []
        (map (fn [h]
               (vec
                (if (and (seq pivot-cols) (not (seq pivot-rows)))
                  (concat (map #(get column-titles %) pivot-cols) h)
                  (concat (map #(get column-titles %) pivot-rows) h)))))
        (let [hs (filter seq column-headers)]
          (when (seq hs)
            (apply map vector hs)))))
```

Reply:
This PR is not only about the performance, but also to demonstrate good hygiene. I agree with you regarding the defensive `vec` calls next to things like `into` and `conj`, but I usually use those when the target doesn't originate in the current function. Here the `into` and the `mapv` are next to each other. If somebody breaks this relation – well, it is partly on them.